pills/20-basic-dependencies-and-hooks.xml

```diff
@@ -0,0 +1,179 @@
+      One can almost think of this as an escape hatch around Nix's normal isolation guarantees, and the principle that dependencies are immutable and inert.
+      We're not actually doing something unsafe or modifying dependencies, but we are allowing arbitrary ad-hoc behavior.
+      For this reason, setup-hooks should only be used as a last resort.
+    </para>
+  </section>
+
+  <section>
+    <title>Environment Hooks</title>
+
+    <para>
+      As a final convenience, we have environment hooks.
+      Recall in <link linkend="inputs-design-pattern">Pill 12</link> how we created <envar>NIX_CFLAGS_COMPILE</envar> for <literal>-I</literal> flags and <envar>NIX_LDFLAGS</envar> for <literal>-L</literal> flags, in a similar manner to how we prepared the <envar>PATH</envar>.
+      One point of ugliness was how anti-modular this was.
+      It makes sense to build the <envar>PATH</envar> in generic builder, because the <envar>PATH</envar> is used by the shell, and the generic builder is intrinsically tied to the shell.
+      But <literal>-I</literal> and <literal>-L</literal> flags are only relevant to the C compiler.
+      The stdenv isn't wedded to including a C compiler (though it does by default), and there are other compilers too which may take completely different flags.
+    </para>
+
+    <para>
+      As a first step, we can move that logic to a setup hook on the C compiler;
+      indeed that's just what we do in CC Wrapper.
+      <footnote>
+        <para>
+          <link xlink:href="https://github.com/NixOS/nixpkgs/tree/6675f0a52c0962042a1000c7f20e887d0d26ae25/pkgs/build-support/gcc-wrapper" />. It was called GCC Wrapper in the version of nixpkgs we're using in this pill; Darwin and Clang support hadn't yet motivated the rename.
+        </para>
+      </footnote>
+      But this pattern comes up fairly often, so somebody decided to add some helper support to reduce boilerplate.
+    </para>
+
+    <para>
+      The other half of <function>addToEnv</function> is:
+      <screen><xi:include href="./20/env-hooks-0.bash" parse="text" /></screen>
+      See how for every package we iterate through the contents of <varname>envHooks</varname>, and apply every function within to the package in question.
+      The idea is one can write a setup hook like:
+      <screen><xi:include href="./20/env-hooks-1.bash" parse="text" /></screen>
+      and if one dependency has that setup hook then all of them will be so <command>echo</command>ed.
+      Allowing dependencies to learn about their <emphasis>sibling</emphasis> dependencies is exactly what compiler need.
+    </para>
+
+  </section>
+
+  <section>
+    <title>Next pill...</title>
+    <para>
+      ...I'm not sure!
+      We could talk the additional dependency types and hooks which cross compilation necessitates, building on our knowledge here to cover stdenv as it works today.
```

Comment:
I propose you for a writeup on cross-compilation: @Ericson2314

Reply:
Hmm. Implementation wise, cross compilation is intentionally part of everything else, and not one distinct add-on. I suppose ideally I'd document each thing with cross compilation support and the overall methodology.